components/ws-proxy/pkg/sshproxy/server.go

```diff
@@ -161,15 +191,38 @@ func (s *Server) HandleConn(c net.Conn) {
 	}
 }
 
-func (s *Server) Authenticator(workspaceId, ownerToken string) (err error) {
+func (s *Server) Authenticator(workspaceId, ownerToken string) (*p.WorkspaceInfo, error) {
 	wsInfo := s.workspaceInfoProvider.WorkspaceInfo(workspaceId)
 	if wsInfo == nil {
-		return fmt.Errorf("not found workspace")
+		return nil, ErrWorkspaceNotFound
 	}
 	if wsInfo.Auth.OwnerToken != ownerToken {
-		return fmt.Errorf("auth failed")
+		return wsInfo, ErrAuthFailed
 	}
-	return nil
+	return wsInfo, nil
+}
+
+func (s *Server) TrackSSHConnection(wsInfo *p.WorkspaceInfo, phase string, err error) {
+	// if we didn't find an associated user, we don't want to track
+	if wsInfo == nil {
+		return
+	}
+	propertics := make(map[string]interface{})
+	propertics["workspaceId"] = wsInfo.WorkspaceID
+	propertics["instanceId"] = wsInfo.InstanceID
+	propertics["state"] = "success"
+	propertics["phase"] = phase
+
+	if err != nil {
+		propertics["state"] = "failed"
+		propertics["cause"] = err.Error()
+	}
+
+	tracker.Track(analytics.TrackMessage{
+		Identity:   analytics.Identity{UserID: wsInfo.OwnerUserId},
+		Event:      "ssh_connection",
+		Properties: propertics,
```

Comment:
Can we provide timestamp here to make sure that events are ordered properly? Otherwise I'm not sure whether they can race in Segment.